### PR TITLE
Close commandline on backspace / when leading prefix gets deleted.

### PR DESCRIPTION
**Pull request by Argn0**
closes https://github.com/qutebrowser/qutebrowser/issues/3458

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/3478)

### CODE REVIEW

File: qutebrowser/config/configdata.yml [hunk 1]
```diff
@@ -740,6 +740,11 @@ completion.use_best_match:
   default: false
   desc: Execute the best-matching command on a partial match.
 
+completion.exit_prefix:
```
Comment (The-Compiler) [inline qutebrowser/config/configdata.yml]: Do you think it really makes sense to add a setting for this?

File: qutebrowser/config/configdata.yml [hunk 2]
```diff
@@ -2493,4 +2498,4 @@ bindings.commands:
       keyboard.
 
     * register: Entered when qutebrowser is waiting for a register name/key for
-      commands like `:set-mark`.
+      commands like `:set-mark`.
```
Comment (The-Compiler) [inline qutebrowser/config/configdata.yml]: Unrelated whitespace change.

File: qutebrowser/mainwindow/statusbar/command.py [hunk 3]
```diff
@@ -31,7 +31,6 @@
 
 
 class Command(misc.MinimalLineEditMixin, misc.CommandLineEdit):
-
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/command.py]: Unrelated whitespace change.

File: qutebrowser/mainwindow/statusbar/command.py [hunk 4]
```diff
@@ -55,6 +54,8 @@ class Command(misc.MinimalLineEditMixin, misc.CommandLineEdit):
     show_cmd = pyqtSignal()
     hide_cmd = pyqtSignal()
 
+    prefixes = '/ : ?'
+
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/command.py]: You can just use `modeparsers.STARTCHARS`.

File: qutebrowser/mainwindow/statusbar/command.py [hunk 5]
```diff
@@ -68,6 +69,7 @@ def __init__(self, *, win_id, private, parent=None):
         self.textChanged.connect(self.update_completion)
         self.textChanged.connect(self.updateGeometry)
         self.textChanged.connect(self._incremental_search)
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/command.py]: Did you verify that all those methods can deal with an empty text?

File: qutebrowser/mainwindow/statusbar/command.py [hunk 6]
```diff
@@ -259,3 +264,9 @@ def _incremental_search(self, text):
 
         if self.prefix() in ['/', '?']:
             self.got_cmd[str].emit(search_prefixes[text[0]] + text[1:])
+
+    @pyqtSlot(str)
+    def _exit_prefix(self, text):
+        if config.val.completion.exit_prefix and (not text or text[0] not in self.prefixes):
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/command.py]: How can `text[0] not in self.prefixes` happen? Wouldn't it always be completely empty?

File: qutebrowser/mainwindow/statusbar/command.py [hunk 7]
```diff
@@ -259,3 +264,9 @@ def _incremental_search(self, text):
...
+            modeman.leave(self._win_id, usertypes.KeyMode.command,
+                          'prefix deleted', maybe=True)
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/command.py]: Why `maybe=True`? If this gets called, we're guaranteed to be in command mode, right?

File: qutebrowser/mainwindow/statusbar/command.py [hunk 8]
```diff
@@ -219,7 +221,10 @@ def setText(self, text):
         if not text:
             pass
         elif text[0] in modeparsers.STARTCHARS:
-            super().set_prompt(text[0])
+            if config.val.completion.exit_prefix:
+                super().set_prompt('')
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/command.py]: This doesn't work correctly - for example, pressing `<Home>` now goes to the beginning of the line. Instead, you can probably remove `miscwidgets._CommandValidator`.

File: qutebrowser/mainwindow/statusbar/command.py (hunk 7 above)
Comment (Argn0) [reply qutebrowser/mainwindow/statusbar/command.py]: Without this it crashes after executing a command.

File: qutebrowser/mainwindow/statusbar/command.py (hunk 5 above)
Comment (Argn0) [reply qutebrowser/mainwindow/statusbar/command.py]: I didn't find any issues.

File: qutebrowser/mainwindow/statusbar/command.py (hunk 6 above)
Comment (Argn0) [reply qutebrowser/mainwindow/statusbar/command.py]: In case the user only deletes the prefix and leaves the rest of the command.

File: qutebrowser/mainwindow/statusbar/command.py [hunk 9]
```diff
@@ -215,12 +216,8 @@ def on_mode_left(self, mode):
...
 
     def setText(self, text):
-        """Extend setText to set prefix and make sure the prompt is ok."""
-        if not text:
-            pass
-        elif text[0] in modeparsers.STARTCHARS:
-            super().set_prompt(text[0])
-        else:
+        """Extend setText to make sure the prefix is valid."""
+        if text and text[0] not in modeparsers.STARTCHARS:
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/command.py]: This still breaks things like `<Home>` :wink:

File: qutebrowser/mainwindow/statusbar/command.py [hunk 10]
```diff
@@ -259,3 +256,9 @@ def _incremental_search(self, text):
 
         if self.prefix() in ['/', '?']:
             self.got_cmd[str].emit(search_prefixes[text[0]] + text[1:])
+
+    @pyqtSlot(str)
+    def _exit_prefix(self, text):
+        if not text or text[0] not in modeparsers.STARTCHARS:
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/command.py]: I don't think we should exit command mode when deleting the prefix but there's still text - vim, for example, doesn't either (it just disallows deleting it).

File: qutebrowser/mainwindow/statusbar/command.py (hunk 9 above)
Comment (Argn0) [reply qutebrowser/mainwindow/statusbar/command.py]: How can I reproduce this problem? When I'm in command mode I can't even press Home.

Comment (Argn0) [reply qutebrowser/mainwindow/statusbar/command.py]: Or do you mean the home method ?

**Comment by Argn0** [comment]
I found a new way to do this. <Home> is working correctly now and the behavior is like in VIM.

File: qutebrowser/mainwindow/statusbar/command.py [hunk 11]
```diff
@@ -232,6 +232,11 @@ def keyPressEvent(self, e):
         Enter/Shift+Enter/etc. will cause QLineEdit to think it's finished
         without command_accept to be called.
         """
+        text = self.text()
+        if text in modeparsers.STARTCHARS and e.key() == Qt.Key_Backspace:
+            modeman.leave(self._win_id, usertypes.KeyMode.command,
+                          'prefix deleted')
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/statusbar/command.py]: You should also call `e.accept()` and `return` so the key press isn't processed further (as we're leaving insert mode anyways).

**Comment by The-Compiler** [comment]
Definitely much simpler that way - thanks!
